mennekes/integrationpluginmennekes.cpp

```diff
@@ -0,0 +1,522 @@
+                    info->finish(Thing::ThingErrorNoError);
+                } else {
+                    info->finish(Thing::ThingErrorHardwareFailure);
+                }
+            });
+        }
+    }
+}
+
+void IntegrationPluginMennekes::thingRemoved(Thing *thing)
+{
+    if (thing->thingClassId() == amtronECUThingClassId && m_amtronECUConnections.contains(thing)) {
+        AmtronECUModbusTcpConnection *connection = m_amtronECUConnections.take(thing);
+        delete connection;
+    }
+
+    if (thing->thingClassId() == amtronHCC3ThingClassId && m_amtronHCC3Connections.contains(thing)) {
+        AmtronHCC3ModbusTcpConnection *connection = m_amtronHCC3Connections.take(thing);
+        delete connection;
+    }
+
+    // Unregister related hardware resources
+    if (m_monitors.contains(thing))
+        hardwareManager()->networkDeviceDiscovery()->unregisterMonitor(m_monitors.take(thing));
+
+    if (myThings().isEmpty() && m_pluginTimer) {
+        hardwareManager()->pluginTimerManager()->unregisterTimer(m_pluginTimer);
+        m_pluginTimer = nullptr;
+    }
+}
+
+void IntegrationPluginMennekes::updateECUPhaseCount(Thing *thing)
+{
+    AmtronECUModbusTcpConnection *amtronECUConnection = m_amtronECUConnections.value(thing);
+    int phaseCount = 0;
+    qCDebug(dcMennekes()) << "Phases: L1" << amtronECUConnection->meterCurrentL1() << "L2" << amtronECUConnection->meterCurrentL2() << "L3" << amtronECUConnection->meterCurrentL3();
+    // the current idles on some 5 - 10 mA when not charging...
+    // We want to detect the phases we're actually charging on. Checking the current flow for that if it's > 500mA
+    // If no phase is charging, let's count all phases that are not 0 instead (to determine how many phases are connected at the wallbox)
+
+    if (amtronECUConnection->meterCurrentL1() > 500) {
+        phaseCount++;
+    }
+    if (amtronECUConnection->meterCurrentL2() > 500) {
+        phaseCount++;
+    }
+    if (amtronECUConnection->meterCurrentL3() > 500) {
+        phaseCount++;
+    }
+    qCDebug(dcMennekes()) << "Actively charging phases:" << phaseCount;
+    if (phaseCount == 0) {
+        if (amtronECUConnection->meterCurrentL1() > 0) {
+            phaseCount++;
+        }
+        if (amtronECUConnection->meterCurrentL2() > 0) {
+            phaseCount++;
+        }
+        if (amtronECUConnection->meterCurrentL3() > 0) {
+            phaseCount++;
+        }
+        qCDebug(dcMennekes()) << "Connected phases:" << phaseCount;
+    }
+
+    thing->setStateValue(amtronECUPhaseCountStateTypeId, phaseCount);
+}
+
+void IntegrationPluginMennekes::setupAmtronECUConnection(ThingSetupInfo *info)
+{
+    Thing *thing = info->thing();
+
+    QHostAddress address = m_monitors.value(thing)->networkDeviceInfo().address();
+
+    qCDebug(dcMennekes()) << "Setting up amtron wallbox on" << address.toString();
+    AmtronECUModbusTcpConnection *amtronECUConnection = new AmtronECUModbusTcpConnection(address, 502, 0xff, this);
```

Comment:
No, the slave address is not configurable. The thing that is configurable though is the port... Mennekes seems to abuse the port for the modbus slave id... However, for the models we support I see absolutely no reason for the user to change the port (it's not like he'd want to use the port for something else on the wallbox) and given we'd already need the port as a discovery param this is IMO too complex to handle.

There is, however, one Mennekes charging pole with multiple controllers on the same network host. In which case the second controller would have the port 503 instead of 503. I would, not implement that until we actually have a use case and such a device for testing it.

=> so bottomline, I'd currently say we only support the default port configuration for now.